### ISSUE TITLE
Typhoeus causes CPU to max out when a large volume of requests are sent to an endpoint that times out, preventing requests to other endpoints from being sent.

### ISSUE THREAD

**Issue by jonnyom** [typhoeus/typhoeus#618]
Hi there, we’ve been using Typhoeus internally in Intercom for some time now and for the most part, it’s been great.

In a recent re-write of our webhooks service we placed Typhoeus in a central part of the service as the HTTP client. Due to the nature of the service,  we can’t guarantee that every endpoint we send requests to will respond in a timely manner, or adhere to our 5 second timeout. 

Having been running the service in production for a few months, we began noticing some unusual behaviour with the service. Whenever a spike in errors were seen in the system (if someone takes an action that triggers a large volume of webhook notifications that are being sent to a server that cannot process them all, which then begins timing out) requests to other, unrelated endpoints began returning timeouts and not delivering as well.

After digging into this internally, we managed to pinpoint the root cause as being Typhoeus.

We can see on our system that there is a gradual degradation to the boxes that are processing the requests.
![image](https://user-images.githubusercontent.com/16747959/51046918-3d20b000-15bf-11e9-887d-16fdc3b81abf.png)

The failures begin trending upwards gradually over time, and then each time processes are restarted on a deploy (the red lines) the error rates drop back to normal.

Our system is based on a series of SQS queues in AWS. Our webhook system picks up notifications off the queue and then delivers them sequentially, process by process (we are *not* using threading at all in this system).

To replicate this we sent a large number of dummy requests through  the pipeline on a specific box to two URLs, one that instantly returned 200 and one that immediately sleeps for 6 seconds.
After about 45 minutes and approximately 100000 requests the CPU on the box spiked to 100% and all requests began returning the libcurl return code `:operation_timedout`, and neither endpoint received any more requests from the system.

![image](https://user-images.githubusercontent.com/16747959/51046933-4742ae80-15bf-11e9-808d-1391e3042c95.png)

Running htop on the instance demonstrates that each of the processes on the box are completely used up.

![image](https://user-images.githubusercontent.com/16747959/51046946-4dd12600-15bf-11e9-9eb5-0bd37422136e.png)

The issue becomes even clearer when looking at our graphing tools

![image](https://user-images.githubusercontent.com/16747959/51046956-54f83400-15bf-11e9-84ab-0ce40451983f.png)

You can see that almost as soon as `good-url` begins seeing `operation_timedout` the `:ok` response code completely drops off.
To remedy this in our test environment we purged the SQS queue and began enqueueing jobs again, but quickly the degradation happened again.

We then began testing with Faraday using the httpclient adapter to see if there would be any difference.
We saw an immediate improvement and no longer saw the CPU spikes or drop in notifications to the good URL.

![image](https://user-images.githubusercontent.com/16747959/51046967-5aee1500-15bf-11e9-8c10-ad5062efc824.png)

![image](https://user-images.githubusercontent.com/16747959/51046981-617c8c80-15bf-11e9-93da-9ae19cd96365.png)

There were still some `operation_timedout` responses from the server when using Faraday, but it looks like those were legitimate as it was just a laptop running the server and the successful requests continued.

Replicating this might be tough. But our exact steps to replicate were as follows:
* Set up two URLS, a “good” URL and a “bad” URL, one that returns 200 and one that times out.
* Using SQS, begin enqueuing jobs, which randomly pick either the good URL or the bad URL, to a worker server that picks up those jobs and begins delivering to the two URLs.
* After about 45 minutes and approximately 100000 requests (combined between good and bad URLs and across each of the processes on the box) the CPU on the box should max out and the processes stop delivering requests

Reproducing locally is taking some time so I can’t reliably say that it happens, but we’ve reproduced it enough with the above steps to be sure that Typhoeus is the cause.

Our Typhoeus configuration looks like this
```
args = {
  method: :post,
  forbid_reuse: true,
  timeout: 5,
  connecttimeout: 1,
  headers: {
    "Content-Type" => "application/json",
    "Accept" => "application/json",
  },
  body: payload,
}

Typhoeus::Request.new(url, args).run
```

Gem version 1.3.1

TL;DR Typhoeus seems to cause CPU spikes and stops processing requests when a large volume of requests time out.

**Comment by uberllama** [comment]
Did you manage to resolve this?

**Comment by jonnyom** [comment]
No, we migrated to Faraday as our HTTP client and haven't seen any further issues.

**Comment by uberllama** [comment]
> No, we migrated to Faraday as our HTTP client and haven't seen any further issues.

👍  Are you still using the Typhoeus adapter, or another underlying lib?

**Comment by jonnyom** [comment]
Using a different adapter, we're now using httpclient!

**Comment by felipedmesquita** [comment]
Hi everyone! Thanks, @jonnyom, for your report.
Since this issue seems resolved (unfortunately by not using our gem) and hasn't had any activity in over five years, I'm closing it. I hope performance improvements in ruby/ffi/ethon have made the situation better today, if anyone is still running into this on current versions and os, please open a new issue and we'll investigate!
